content/umdctf_2024/ready_aim_fire.md

```diff
@@ -0,0 +1,161 @@
+---
+title: ready-aim-fire
+date: 2024-04-26T14:23:28+03:00
+description: Writeup for ready-aim-fire [UMD 2024]
+author: PineBel
+tags:
+  - pwn
+draft: false
+---
+## Challenge Description
+
+Firing your weapon when the spice harvester's shields are down requires exceptional timing.
+
+### Intuition
+This was a fun challenge which I solved in an unintented way, I will also present the intented solution because it's very interesting. 
+We get the source code (and binary) for this challenge and we can see that we have a BOF in the fire method from the Cannon object and we also have a stack leak.
+
+```C
+void print_flag() {
+    ifstream f{"./flag.txt"};
+    if (!f.is_open()) {
+        cout << "Failed to open flag file. Contact CTF organizers if you see this error." << endl;
+    } else {
+        string flag;
+        f >> flag;
+        cout << flag << endl;
+    }
+}
+
+
+void direct_hit() {
+    try {
+        throw exception{};
+    } catch (exception e) {
+        cout << "Direct hit!" << endl;
+        print_flag();
+    }
+}
+
+class Cannon {
+public:
+    int bufIndex;
+    char buf[32];
+
+    Cannon(): bufIndex(0) {}
+
+    void fire() {
+        char c ;
+        for (;;) {
+            cin.get(c);
+            if (c == '\n') {
+                break;
+            } else {
+                buf[bufIndex++] = c;
+            }
+        }
+        if (bufIndex >= 32) {
+            throw out_of_range{""};
+        }
+    }
+};
+
+void fire_weapon() {
+    Cannon w;
+    w.fire();
+}
+
+int main() {
+    int target_assist;
+    cout << "Quick! While the spice harvester's shields are down! Fire the laser cannon!" << endl;
+    cout << &target_assist << endl;
+
+    try {
+        fire_weapon();
+        cout << "Looks like you missed your opportunity to fire." << endl;
+    }
+    catch (exception e) {
+        cout << "Seems like you missed." << endl;
+    }
+}
+```
+
+We can see that there is a constraint though, if we do a BOF, it will trigger an exception, which will be caught in the main function. Additionally, I noticed an extra function called 'direct_hit,' which suggested that we should attempt to redirect the exception thrown during a BOF into the catch block of 'direct_hit()' to print the flag (the intended solution). Unfortunately, my implementation didn't work as expected. Therefore, my next idea was to overwrite the return address of the main function with the address of 'print_flag()' after the exception was triggered
+
+### Solution
+
+#### Unintented
+
+When first trying to do the BOF the program crashed in the exception handler and we could clearly see that we had overwritten there something.
```

Comment:
To me it seems like you tend to have sentences on the longer(ish) side, which makes everything harder to read imo. For example here I would rephrase like:

When first trying to do the BOF the program crashed in the exception handler. We could clearly see that we had overwritten something there.

This change would also give you a bit more space to provide extra details: 
* BOF with what input?
* _could clearly see_ -- q: what where? | a: GDB i guess? 
* _overwritten something_ -- what is something?